src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,184 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node newNode = new Node<>(value);
+        if (size == 0) {
+            head = tail = newNode;
+            size++;
+            return;
+        }
+        newNode.previous = tail;
+        tail.next = newNode;
+        tail = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == 0) {
+            addAsHead(value);
+            return;
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkIndex(index);
+        Node newNode = new Node(value);
+        Node currentNode = getNodeByIndex(index);
+        insertBefore(currentNode, newNode);
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        if (index == 0) {
+            return head.value;
+        }
+        if (index == size - 1) {
+            return tail.value;
+        }
+        return (T) getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        T replacedValue;
+        if (index == 0) {
+            replacedValue = head.value;
+            head.value = value;
+            return replacedValue;
+        }
+        if (index == size - 1 && index > 0) {
+            replacedValue = tail.value;
+            tail.value = value;
+            return replacedValue;
+        }
+        checkIndex(index);
+        Node current = getNodeByIndex(index);
+        replacedValue = (T) current.value;
+        current.value = value;
+        return replacedValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node removedNode;
+        if (size == 1 && index == 0) {
+            removedNode = head;
+            head = tail = null;
+            size--;
+            return (T) removedNode.value;
+        }
+        if (index == 0) {
+            removedNode = head;
+            head = removedNode.next;
+            head.previous = null;
+            size--;
+            return (T) removedNode.value;
+        }
+        checkIndex(index);
+        if (index == size - 1) {
+            removedNode = tail;
+            tail = removedNode.previous;
+            tail.next = null;
+            size--;
+            return (T) removedNode.value;
+        }
```

Comment:
![image](https://user-images.githubusercontent.com/81781479/185737804-fec659d9-db6d-4f27-835c-dcb6b7f5f630.png)

move this to method unlink